src/rt/ds/stack.h

```diff
@@ -7,128 +7,302 @@
+      auto next = get_block(iter);
+
+      assert(index != (T**)&null_block);
+      next->prev = index;
+      index = &(next->data[0]);
+      next->data[0] = item;
+    }
+
+    /// For all elements of the stack
+    template<void apply(T* t)>
     void forall()
     {
-      Block* curr = block;
-      size_t i = index;
+      T* curr = index;
 
-      while (curr != nullptr)
+      while (curr != null_index)
       {
         do
         {
-          i--;
-          apply(curr->data[i]);
-        } while (i > 0);
+          apply(*curr);
+          curr--;
+        } while (is_empty(curr));
 
-        curr = curr->prev;
-        i = STACK_COUNT;
+        curr = get_block(curr)->prev;
       }
     }
+  };
 
-  private:
-    void pop_slow_path()
+  /**
+   * This class uses the block structured stack with extra fields
+   * for the allocator, and a backup block, so that the common case
+   * of 0-1 elements can be fast, and any other block boundrary case.
+   */
+  template<class T, class Alloc>
+  class Stack
+  {
+    using Block = typename StackBase<T, Alloc>::Block;
+    StackBase<T, Alloc> stack;
+    Alloc* alloc;
+    Block* backup;
+
+  public:
+    Stack(Alloc* alloc) : alloc(alloc), backup(nullptr) {}
+
+    ALWAYSINLINE void push(T* item)
     {
-      Block* prev = block->prev;
+      if (stack.push_is_fast())
+      {
+        stack.push_fast(item);
+        return;
+      }
 
-      if (backup != nullptr)
-        alloc->template dealloc<sizeof(Block)>(backup);
+      Block* new_block = backup;
+
+      if (new_block == nullptr)
+        new_block = (Block*)alloc->template alloc<sizeof(Block)>();
+
+      backup = nullptr;
 
-      backup = block;
-      block = prev;
-      index = STACK_COUNT;
+      stack.push_slow(item, new_block);
     }
 
-    void push_slow_path(T item)
+    T* pop()
     {
-      Block* next;
+      if (stack.pop_is_fast())
+      {
+        return stack.pop_fast();
+      }
 
-      if (backup != nullptr)
+      auto res_block = stack.pop_slow();
+      if (backup == nullptr)
       {
-        next = backup;
-        backup = nullptr;
+        backup = res_block.second;
       }
       else
       {
-        next = (Block*)alloc->template alloc<sizeof(Block)>();
+        alloc->template dealloc<sizeof(Block)>(res_block.second);
       }
+      return res_block.first;
+    }
 
-      index = 1;
-      next->data[0] = item;
-      next->prev = block;
-      block = next;
+    ~Stack()
+    {
+      stack.dealloc(alloc);
+      if (backup != nullptr)
+      {
+        alloc->template dealloc<sizeof(Block)>(backup);
+      }
+    }
+
+    bool empty()
+    {
+      return stack.empty();
+    }
+
+    T* peek()
+    {
+      return stack.peek();
+    }
+  };
+
+  /**
+   * Block structured stack, that is a single pointer in size.
+   *
+   * Operations require an explicit Alloc parameter in case more/less is
+   * required.
+   */
+  template<class T, class Alloc>
+  class StackSmall
```

Comment:
`SmallStack`?

Reply:
More bikeshedding, I'm not a fan of "small" because it reminds me of the `SmallVec`/`SmallSet` found in LLVM and other projects, but small has a completely different meaning there.

How about "ThinStack", from the thin/fat pointer terminology?

Reply:
Shall I create a Doodle Pool? :-)

I think `ThinStack` is good.  I didn't like `StackSmall`, so thank you both for the bike shedding.

Reply:
I was planning on raising this in the context of the standard library, because at some point we will need to define naming conventions for that.  My suggestion would be to use suffixing rather than prefixing because it plays nicer with auto-complete.  Any identifier that can be used in an arbitrary context should start with the noun and add adjectives.  Apple introduced this rule about 15 years ago and renamed all of the OpenStep enums from things like `NSBoldTextFormat` to `NSTextFormatBold` but they didn't do it with classes.  This means you can type `NSTextFormat` and get autocompletions of all of the kinds of text format but you can't type `NSArray` and get all of the completions of class types.  Here, you can type `Stack` and get completions of the different kinds of stack, which is a bit more human friendly.  It's not useful for things that you'll use in a context where an editor (or, rather, LSP implementation) can refine the list of possible identifiers from some type information, but when naming types that isn't generally possible.

Reply:
A lot of editors use substring rather than prefix to filter, then you get nice behaviour, but I am happy to adopt this kind of convention for other editors